### PR TITLE
improve(benchmarks): support multiple values + visualizing runtimes

### PR DESCRIPTION
**Pull request by 0xNineteen**
- docs in `docs/benchmarks.md`
- scripts for visualizing `scripts/view_bench.py`
- completely new benchmarking output from `src/benchmarks.zig`
- add new tests to benchmarks
  - fixes to leaks throughout the benchmarks (some require larger changes -- left for another PR)
- changes to snapshot downloading (from previous PR/rex) 
  - closes account files after mmaping their data in - reducing open file limit errors (data is never modified so its ok) - this also lets us map it in with the PRIVATE flag
  - reduce the accounts-db-snapshot benchmarks to only a single method which benchmarks loading *and* validation in one go 
  - improve snapshot download code
- support multiple values from a benchmark
- visualizing all runtimes (outputted to results/ folder since its not visible)

example output from python script: 
```
Viewing ['b_results.txt', 'b_results2.txt']
Saved to results/readWriteAccounts(100k accounts (1_slot - ram index - ram accounts)) (read_time).png
Saved to results/readWriteAccounts(100k accounts (1_slot - ram index - ram accounts)) (write_time).png
Saved to results/readWriteAccounts(100k accounts (1_slot - disk index - ram accounts)) (read_time).png
Saved to results/readWriteAccounts(100k accounts (1_slot - disk index - ram accounts)) (write_time).png
...
```
readWriteAccounts(100k accounts (1_slot - ram index - ram accounts)) (write_time)
- each point on y-axis=0 is a runtime 
- the point on y-axis=1 is the mean with the bar surrounding it being the standard deviation
---

_note_: builds off #289 -- i switched to mainly working on this feat so moved it to my own branch (instead of rexi's)

### CODE REVIEW

File: docs/benchmarks.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#### visualizing
+
+```bash
+./zig-out/bin/benchmark accounts_db_readwrite -r 2>&1 | tee bench_results.txt # save output to file
+# NOTE: need to format doc to below
+python scripts/view_bench.py bench_results.txt # view runtimes as a charts with one file source
+python scripts/view_bench.py bench_results.txt b_results2.txt # compare runtimes against two *equivalent* files
+```
+
```
Comment (Sobeston) [inline docs/benchmarks.md:33]: Just went to run this example and got a type error, leaving me without any data (?)

`TypeError: unsupported operand type(s) for /: 'str' and 'int'`

image

File: scripts/view_bench.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        plt.clf()
+        plt.title(benchmark_names[i] + f"{units}", wrap=True)
+
+        for df_i, df in enumerate(dfs):
+            benchmark_runtimes = df.T[1:][i]
+            # convert to milliseconds 
+            if units == 'ms':
+                benchmark_runtimes = benchmark_runtimes / 1_000_000
+            if units == 's':
+                benchmark_runtimes = benchmark_runtimes / 1_000_000_000
```
Comment (Sobeston) [inline scripts/view_bench.py:41]: this seems to be the source of the error in my previous comment
```
    benchmark_runtimes = benchmark_runtimes / 1_000_000
                         ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
```

File: src/accountsdb/accounts_file.zig [hunk 3]
```diff
@@ -253,7 +252,7 @@ pub const AccountFile = struct {
             null,
             file_size,
             std.posix.PROT.READ | std.posix.PROT.WRITE,
-            std.posix.MAP{ .TYPE = .SHARED },
+            std.posix.MAP{ .TYPE = .PRIVATE },
             file.handle,
```
Comment (Sobeston) [inline src/accountsdb/accounts_file.zig:256]: Doesn't this mean that writes aren't carried back to the file; is this intentional?

File: src/accountsdb/db.zig [hunk 4]
```diff
@@ -3881,28 +3882,26 @@ pub const BenchmarkAccountsDBSnapshotLoad = struct {
...
-            .allocator = allocator,
-            .max_level = Level.debug,
-            .max_buffer = 1 << 20,
-        });
-        defer std_logger.deinit();
-
-        const logger = std_logger.logger();
+        const logger = Logger{ .noop = {} };
+        // defer logger.deinit();
+        // logger.spawn();
```
Comment (Sobeston) [inline src/accountsdb/db.zig:3904]: dangling comments

File: src/accountsdb/swiss_map.zig [hunk 5]
```diff
@@ -674,28 +684,22 @@ pub const BenchmarkSwissMap = struct {
...
         const read_speedup = @as(f32, @floatFromInt(std_read_time.asNanos())) / @as(f32, @floatFromInt(read_time.asNanos()));
-        const read_faster_or_slower = if (read_speedup < 1.0) "slower" else "faster";
-        std.debug.print("\tREAD: {} ({d:.2}x {s} than std)\n", .{
-            std.fmt.fmtDuration(read_time.asNanos()),
-            read_speedup,
-            read_faster_or_slower,
-        });
-
-        return write_time;
+        // const read_faster_or_slower = if (read_speedup < 1.0) "slower" else "faster";
```
Comment (Sobeston) [inline src/accountsdb/swiss_map.zig:695]: dangling comment

File: src/gossip/service.zig [hunk 6]
```diff
@@ -3436,3 +3438,21 @@ fn localhostTestContactInfo(id: Pubkey) !ContactInfo {
...
+
+// test "benchmarkGossipService" {
+//     _ = try BenchmarkGossipServiceGeneral.benchmarkGossipService(.{
+//         .message_counts = .{
+//             .n_ping = 10,
+//             .n_push_message = 10,
+//             .n_pull_response = 10,
+//         },
+//     });
+// }
```
Comment (Sobeston) [inline src/gossip/service.zig:3458]: Are we fixing and re-enabling these, or removing these?

**Review by Sobeston — CHANGES_REQUESTED** [review]
Seems good overall, but some issues:

File: src/accountsdb/accounts_file.zig (hunk 3 above)
Comment (Rexicon226) [reply src/accountsdb/accounts_file.zig:256]: See previous discussion on the topic: https://github.com/Syndica/sig/pull/289#discussion_r1786576053
TLDR; this is correct.

Comment (Sobeston) [reply src/accountsdb/accounts_file.zig:256]: Perfect, just double checking :ok_hand:

File: docs/benchmarks.md (hunk 1 above)
Comment (0xNineteen) [reply docs/benchmarks.md:33]: [e713769](https://github.com/Syndica/sig/pull/316/commits/e713769884cd166d46927ce9ddd31f00f7e621ac)

File: scripts/view_bench.py (hunk 2 above)
Comment (0xNineteen) [reply scripts/view_bench.py]: [e713769](https://github.com/Syndica/sig/pull/316/commits/e713769884cd166d46927ce9ddd31f00f7e621ac)

File: src/gossip/service.zig (hunk 6 above)
Comment (0xNineteen) [reply src/gossip/service.zig:3459]: left a comment - right now they leak when run with the testing allocator -- will fix the leak in another PR since its a bit more involved to fix

[bb8257f](https://github.com/Syndica/sig/pull/316/commits/bb8257f935ab417799c53a90ccdd3aa96747ce15)

https://github.com/orgs/Syndica/projects/2/views/10?pane=issue&itemId=84676970

File: src/accountsdb/swiss_map.zig (hunk 5 above)
Comment (0xNineteen) [reply src/accountsdb/swiss_map.zig]: [bb8257f](https://github.com/Syndica/sig/pull/316/commits/bb8257f935ab417799c53a90ccdd3aa96747ce15)

File: src/accountsdb/db.zig (hunk 4 above)
Comment (0xNineteen) [reply src/accountsdb/db.zig]: [bb8257f](https://github.com/Syndica/sig/pull/316/commits/bb8257f935ab417799c53a90ccdd3aa96747ce15)

**Comment by 0xNineteen** [comment]
human format supported

Screenshot 2024-10-28 at 3 02 13 PM

File: scripts/view_bench.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import argparse
+import matplotlib.pyplot as plt
+import numpy as np
+import random
+
+def random_color_generator():
+    r = random.randint(0, 255)
+    g = random.randint(0, 255)
+    b = random.randint(0, 255)
+    return (r / 255, g / 255, b / 255)
```
Comment (Sobeston) [inline scripts/view_bench.py:12]:
```suggestion
def random_color_generator():
    # take [r, g, b] from colormap
    # Dark2 chosen for visibility from
    # https://matplotlib.org/stable/users/explain/colors/colormaps.html#qualitative
    return plt.colormaps["Dark2"](random.random())[:3] 
```

File: src/benchmarks.zig [hunk 8]
```diff
@@ -1,326 +1,388 @@
...
+                        const x: std.MultiArrayList(runtime_type).Field = @enumFromInt(j);
+                        for (runtimes.items(x)) |f_runtime| {
+                            const d = if (f_runtime > f_mean) f_runtime - f_mean else f_mean - f_runtime;
+                            switch (@typeInfo(T)) {
+                                .Float => f_variance = d * d,
+                                else => f_variance +|= d *| d,
+                            }
+                        }
+                        f_variance /= n_iters;
+
```
Comment (Sobeston) [inline src/benchmarks.zig:359]: Is squared variance what we want? It's hard to interpret in the human table. For the CSV, we could leave it as  variance in the csv and square it later.

**Review by Sobeston — CHANGES_REQUESTED** [review]
Seems pretty great overall, I've commented a few nits.

Some general improvements I'd like after trying it out a bit:
- for the human view, I think it would make sense to default to milliseconds
- the table can be a bit wide (>200 cols), enough to throw off my workflow a bit. columns can be shortened significantly
- being able to attach extra formattable data to a run would be useful

e.g. here's what we currently get from `zig build benchmark -Doptimize=ReleaseSafe -- accounts_db_readwrite`:
```
cache hits/misses :   50940/49060   (50.94%) - total cache accesses: 100000
cache hits/misses :   50940/49060   (50.94%) - total cache accesses: 100000
cache hits/misses :   50940/49060   (50.94%) - total cache accesses: 100000
cache hits/misses :   50940/49060   (50.94%) - total cache accesses: 100000
cache hits/misses :   50940/49060   (50.94%) - total cache accesses: 100000
cache hits/misses :   59962/40038   (59.96%) - total cache accesses: 100000
cache hits/misses :   59962/40038   (59.96%) - total cache accesses: 100000
cache hits/misses :   59962/40038   (59.96%) - total cache accesses: 100000
cache hits/misses :   59962/40038   (59.96%) - total cache accesses: 100000
cache hits/misses :   59962/40038   (59.96%) - total cache accesses: 100000
cache hits/misses :   88253/11747   (88.25%) - total cache accesses: 100000
cache hits/misses :   88253/11747   (88.25%) - total cache accesses: 100000
cache hits/misses :   88253/11747   (88.25%) - total cache accesses: 100000
cache hits/misses :   88253/11747   (88.25%) - total cache accesses: 100000
cache hits/misses :   88253/11747   (88.25%) - total cache accesses: 100000
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| readWriteAccounts                                                     |  read_time_min |  read_time_max |  read_time_mean |  read_time_variance |  write_time_min |  write_time_max |  write_time_mean |  write_time_variance |
+=======================================================================+================+================+=================+=====================+=================+=================+==================+======================+
| 100k accounts (1_slot - ram index - ram accounts - lru disabled)      |  52055439      |  49476105      |  51015898       |  980304199074       |  94068243       |  73089433       |  80092408        |  72059462679766      |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - disk index - ram accounts - lru disabled)     |  62044868      |  50888886      |  53760871       |  17466527927964     |  84230335       |  78045886       |  80702332        |  4552178788900       |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - disk index - disk accounts - lru disabled)    |  53628711      |  52291895      |  52881811       |  289978477739       |  49406430       |  45870428       |  47920674        |  1542221031515       |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - ram index - disk accounts - lru disabled)     |  55502624      |  52783879      |  53529521       |  1031699069252      |  43759489       |  40716693       |  42425920        |  1022610174131       |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - ram index - disk accounts - lru_size=1_000)   |  81294636      |  69163413      |  71876892       |  22337507238103     |  43480649       |  40797218       |  42132997        |  1009236851442       |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - ram index - disk accounts - lru_size=10_000)  |  85961677      |  69886795      |  74741912       |  34340440024912     |  43236761       |  42631969       |  42952303        |  48122424238         |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
| 100k accounts (1_slot - ram index - disk accounts - lru_size=100_000) |  74876549      |  65345722      |  67721228       |  12929284935628     |  46386455       |  42865655       |  44073173        |  1482138470335       |
+-----------------------------------------------------------------------+----------------+----------------+-----------------+---------------------+-----------------+-----------------+------------------+----------------------+
```

Versus what it could potentially look like:

```
+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| readWriteAccounts                                                     |        read_time (ms)         |     write_time (ms)           | cache_hit_rate |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                       | min  | max  | mean | variance | min  | max  | mean | variance |   (custom)     |
|========================================================================================================================================================|
| 100k accounts (1_slot - ram index - ram accounts - lru disabled)      | 52.1 | 49.5 | 51.0 | 0.99     | 94.1 | 73.1 | 80.1 | 8.45     | null           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - disk index - ram accounts - lru disabled)     | .... | .... | .... | ....     | .... | .... | .... | ....     | null           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - disk index - disk accounts - lru disabled)    | .... | .... | .... | ....     | .... | .... | .... | ....     | null           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - ram index - disk accounts - lru disabled)     | .... | .... | .... | ....     | .... | .... | .... | ....     | null           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - ram index - disk accounts - lru_size=1_000)   | .... | .... | .... | ....     | .... | .... | .... | ....     | 50.94          |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - ram index - disk accounts - lru_size=10_000)  | .... | .... | .... | ....     | .... | .... | .... | ....     | 59.96          |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 100k accounts (1_slot - ram index - disk accounts - lru_size=100_000) | .... | .... | .... | ....     | .... | .... | .... | ....     | 88.25          |
+--------------------------------------------------------------------------------------------------------------------------------------------------------+
```

File: src/benchmarks.zig (hunk 8 above)
Comment (Rexicon226) [reply src/benchmarks.zig:359]: Maybe I'm misunderstanding something, but there's no squared variance here? `d * d` *is* variance, where `d` is just deviation.
I discussed this with Brennan before, and I think that standard deviation is a better benchmark for us than variance. (sqrt `f_variance` after all of these computations)

File: src/accountsdb/db.zig [hunk 9]
```diff
@@ -3958,7 +3954,11 @@ pub const BenchmarkAccountsDBSnapshotLoad = struct {
         const dir_path = sig.TEST_DATA_DIR ++ "bench_snapshot/";
         var snapshot_dir = std.fs.cwd().openDir(dir_path, .{ .iterate = true }) catch {
             std.debug.print("need to setup a snapshot in {s} for this benchmark...\n", .{dir_path});
-            return sig.time.Duration.fromNanos(0);
+            const zero_duration = sig.time.Duration.fromNanos(0);
+            return .{
+                .load_time = zero_duration.asNanos(),
+                .validate_time = zero_duration.asNanos(),
+            };
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:3961]: this one is very specific and expensive - in the case where you want to run all benchmarks, imo we dont want to error out and stop all of them just because they dont want to run this one

File: src/benchmarks.zig (hunk 8 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: [74d9015](https://github.com/Syndica/sig/pull/316/commits/74d9015aec1985cf2d2d824107ae422aa98b7af5)

File: scripts/view_bench.py (hunk 7 above)
Comment (0xNineteen) [reply scripts/view_bench.py]: [74d9015](https://github.com/Syndica/sig/pull/316/commits/74d9015aec1985cf2d2d824107ae422aa98b7af5)

File: src/accountsdb/download.zig [hunk 10]
```diff
@@ -302,81 +304,93 @@ pub fn downloadSnapshotsFromGossip(
...
+        nmemb: c_uint,
+        user_data: *anyopaque,
+    ) callconv(.C) c_uint {
+        assert(size == 1); // size will always be 1
         const len = size * nmemb;
-        var self: *Self = @alignCast(@ptrCast(user_data));
-        var typed_data: [*]u8 = @ptrCast(ptr);
+        const self: *Self = @alignCast(@ptrCast(user_data));
+        var typed_data: [*]u8 = @ptrCast(ptr.?);
         const buf = typed_data[0..len];
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/download.zig]: [74d9015](https://github.com/Syndica/sig/pull/316/commits/74d9015aec1985cf2d2d824107ae422aa98b7af5)

File: src/accountsdb/download.zig [hunk 11]
```diff
@@ -302,81 +304,93 @@ pub fn downloadSnapshotsFromGossip(
...
-
         return .{
             .logger = logger,
-            .mmap = file_memory,
-            .download_size = download_size,
+            .file = file,
             .min_mb_per_second = min_mb_per_second,
-            .mb_timer = try std.time.Timer.start(),
+            .mb_timer = undefined,
         };
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/download.zig]: changed from undefined to use a helper function -- it will always be used so optional doesnt make sense -- undefined doesnt feel right -- 
[74d9015](https://github.com/Syndica/sig/pull/316/commits/74d9015aec1985cf2d2d824107ae422aa98b7af5)

**Comment by 0xNineteen** [comment]
now supports
- defined time units per benchmark (all use millis, expect the ones which need nanos) 
- slimmer output format

will leave the custom args to another PR if we find we need them more often -- the random print statements arent that bad imo

Screenshot 2024-11-01 at 1 53 49 PM

File: src/benchmarks.zig [hunk 12]
```diff
@@ -1,349 +1,538 @@
...
+pub const BenchTimeUnit = enum {
+    Nanos,
+    Millis,
+    Seconds,
+
+    pub fn convertDuration(self: BenchTimeUnit, duration: Duration) u64 {
+        return switch (self) {
+            .Nanos => duration.asNanos(),
+            .Millis => duration.asMillis(),
+            .Seconds => duration.asSecs(),
```
Comment (InKryption) [inline src/benchmarks.zig:19]: Why the titlecase enum members?

File: src/benchmarks.zig [hunk 13]
```diff
@@ -1,349 +1,538 @@
...
+    while (true) {
+        const maybe_index = std.mem.indexOf(u8, raw_benchmark_name[index + 1 ..], ".");
+        index += 1;
+        index += maybe_index orelse break;
+    }
+    const benchmark_name = raw_benchmark_name[index..];
+    results_dir.makeDir(benchmark_name) catch |err| {
+        if (err == std.fs.Dir.MakeError.PathAlreadyExists) {} else {
+            return err;
         }
```
Comment (InKryption) [inline src/benchmarks.zig:218]:
```suggestion
        switch (err) {
            error.PathAlreadyExists => {},
            else => |e| return e,
        }
```

File: src/benchmarks.zig [hunk 14]
```diff
@@ -1,349 +1,538 @@
...
+                        } else {
+                            // single return type
+                            break :blk switch (@TypeOf(arg)) {
+                                void => .{},
+                                else => .{arg},
+                            };
+                        }
+                    },
+                    else => unreachable,
+                }
```
Comment (InKryption) [inline src/benchmarks.zig:266]:
```suggestion
                // NOTE: to know if we should pass in the time unit we
                // check the input params of the function, so any multi-return
                // function NEEDS to have the time unit as the first parameter
                const info = @typeInto(@TypeOf(benchFunction)).Fn;
                const has_time_unit = info.params.len > 0 and info.params[0].type.? == BenchTimeUNit;
                const time_arg = if (has_time_unit) .{time_unit} else .{};
                const other_arg = if (@TypeOf(arg) != void) .{arg} else .{};
                break :blk time_arg ++ other_arg;

```

File: src/benchmarks.zig [hunk 15]
```diff
@@ -1,349 +1,538 @@
...
+            };
+            const result_type: type = @TypeOf(try @call(.auto, benchFunction, arguments));
```
Comment (InKryption) [inline src/benchmarks.zig:268]: Should be `ResultType`

File: src/benchmarks.zig [hunk 16]
```diff
@@ -1,349 +1,538 @@
...
+                switch (result_type) {
+                    Duration => {
+                        const duration = try @call(.auto, benchFunction, arguments);
+                        const runtime = time_unit.convertDuration(duration);
+                        min = @min(runtime, min);
+                        max = @max(runtime, max);
+                        sum += runtime;
+                        try runtimes.append(allocator, .{ .result = runtime });
+                    },
+                    inline else => {
```
Comment (InKryption) [inline src/benchmarks.zig:313]: Why `inline else`? My intuition is that either this is:
a. harmful because it's generating a branch for every possible type in the program (that seems bad).
b. superfluous because the set of types is already comptime-known, meaning it does nothing.

File: src/benchmarks.zig [hunk 17]
```diff
@@ -1,349 +1,538 @@
...
+                        const d_sq = d *| d;
+                        variance +|= d_sq;
+                    }
+                    variance /= iter_count;
+                    const std_dev = std.math.sqrt(variance);
+
+                    // print column results
+                    try writer_average.print("{s}, {d}, {d}, {d}, {d}\n", .{ arg_name, min, max, mean, std_dev });
+                },
+                inline else => {
```
Comment (InKryption) [inline src/benchmarks.zig:390]: Same thought as other comment on the type `inline else`.

File: src/benchmarks.zig [hunk 18]
```diff
@@ -1,349 +1,538 @@
...
+                        } else {
+                            // single return type
+                            break :blk switch (@TypeOf(args[0])) {
+                                void => .{},
+                                else => .{args[0]},
+                            };
+                        }
+                    },
+                    else => unreachable,
+                }
```
Comment (InKryption) [inline src/benchmarks.zig:481]: Same kind of suggestion as on other `switch (@typeInfo(@TypeOf(benchFunction)))`

File: src/benchmarks.zig [hunk 19]
```diff
@@ -1,349 +1,538 @@
...
+                    Duration => {
+                        break :blk struct { result: u64 };
+                    },
+                    // multiple values
+                    else => {
+                        break :blk result_type;
+                    },
+                }
+            };
+            const U = @typeInfo(runtime_type).Struct;
```
Comment (InKryption) [inline src/benchmarks.zig:496]: Should be a lower/snakecase name, it isn't a type. Also would benefit from a nicer name, like `struct_info`, or even just `info`.

File: src/utils/fmt.zig [hunk 20]
```diff
@@ -188,19 +188,19 @@ pub fn newlinesToSpaces(comptime str: []const u8) [str.len]u8 {
 /// Tries to format the real path resolved from `dir` and `pathname`.
 /// Should it encounter an error when doing so, `"(error.Name)/pathname"`
 /// is printed instead.
-pub inline fn tryRealPath(dir: std.fs.Dir, pathname: []const u8) TryRealPathFmt {
-    return .{
+pub inline fn tryRealPath(dir: std.fs.Dir, pathname: []const u8) std.fmt.Formatter(TryRealPathCtx.realPathFormat) {
```
Comment (InKryption) [inline src/utils/fmt.zig:191]: There's no point in using `std.fmt.Formatter` (I think it should be removed altogether). Any reason you did this?

**Review by InKryption — CHANGES_REQUESTED** [review]
Overall good changes, great to see benchmarking evolve in this way + some good improvements in all the adjacent code. Just some comments, suggestions, and doubts I want to address.

File: src/accountsdb/db.zig [hunk 21]
```diff
@@ -4339,7 +4354,8 @@ pub const BenchmarkAccountsDB = struct {
             var i: usize = 0;
             while (i < n_accounts) : (i += 1) {
                 const pubkey_idx = indexer.sample();
-                _ = try accounts_db.getAccount(&pubkeys[pubkey_idx]);
+                const account = try accounts_db.getAccount(&pubkeys[pubkey_idx]);
+                account.deinit(allocator);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:4358]: get account clones the underlying data

[the remark this answers is not in the document]
Comment (InKryption) [reply src/accountsdb/db.zig:4358]: ahh, I forgot about that, apologies.

File: src/accountsdb/db.zig [hunk 22]
```diff
@@ -502,13 +500,16 @@ pub const AccountsDB = struct {
...
+                        try accounts_dir.realpathAlloc(self.allocator, file_name_bounded.constSlice()),
+                        @errorName(err),
+                    });
                     return err;
                 };
-                errdefer accounts_file_file.close();
+                defer accounts_file.close();
 
-                break :blk AccountFile.init(accounts_file_file, file_info, slot) catch |err| {
+                break :blk AccountFile.init(accounts_file, file_info, slot) catch |err| {
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:512]: the file isnt continued to be help, just the memory mapped part of it -- See previous discussion on the topic: https://github.com/Syndica/sig/pull/289#discussion_r1786576053

File: src/accountsdb/db.zig [hunk 23]
```diff
@@ -1742,6 +1742,7 @@ pub const AccountsDB = struct {
                 accounts_alive_size,
                 slot,
             );
+            defer new_file.close();
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:1745]: will resolve these and keep the main one open

File: src/accountsdb/db.zig (hunk 22 above)
[the remark this answers is not in the document]
Comment (InKryption) [reply src/accountsdb/db.zig:512]: This isn't related to that, just the fact that we're storing a file handle that is invalidated after this point. the `accounts_file` is passed as a parameter to initialize the `AccountsFile`, and then the `defer accounts_file.close()` runs.

File: src/accountsdb/db.zig [hunk 24]
```diff
@@ -4185,22 +4196,23 @@ pub const BenchmarkAccountsDB = struct {
...
         const slot_list_len = bench_args.slot_list_len;
         const total_n_accounts = n_accounts * slot_list_len;
 
-        var allocator = std.heap.c_allocator;
-
+        // make sure theres no leaks
+        const allocator = if (builtin.is_test) std.testing.allocator else std.heap.c_allocator;
         const disk_path = sig.TEST_DATA_DIR ++ "tmp/";
         std.fs.cwd().makeDir(disk_path) catch {};
+        defer std.fs.cwd().deleteTreeMinStackSize(disk_path) catch {};
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig]: [eb404ec](https://github.com/Syndica/sig/pull/316/commits/eb404ec1155fc2d7b3a8d6926e8c933daf466881)

File: src/accountsdb/db.zig [hunk 25]
```diff
@@ -4222,12 +4234,17 @@ pub const BenchmarkAccountsDB = struct {
                 std.fs.cwd().deleteFile(filepath) catch {
                     std.debug.print("failed to delete file: {s}\n", .{filepath});
                 };
+                allocator.free(filepath);
             }
         }
 
-        if (bench_args.accounts == .ram) {
+        const write_time = if (bench_args.accounts == .ram) timer_blk: {
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig]: [eb404ec](https://github.com/Syndica/sig/pull/316/commits/eb404ec1155fc2d7b3a8d6926e8c933daf466881)

File: src/accountsdb/download.zig [hunk 26]
```diff
@@ -11,7 +11,9 @@ const ThreadSafeContactInfo = sig.gossip.data.ThreadSafeContactInfo;
 const GossipService = sig.gossip.GossipService;
 const Logger = sig.trace.Logger;
 
-const DOWNLOAD_PROGRESS_UPDATES_NS = 30 * std.time.ns_per_s;
+const assert = std.debug.assert;
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/download.zig]: removed most - [1764858](https://github.com/Syndica/sig/pull/316/commits/1764858a8e688817da750f1e811960118c44e62b) - some the type alias makes sense because its used so much throughout the code

File: src/accountsdb/swiss_map.zig [hunk 27]
```diff
@@ -608,21 +610,21 @@ fn generateData(allocator: std.mem.Allocator, n_accounts: usize) !struct {
     var prng = std.Random.DefaultPrng.init(0);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/swiss_map.zig:610]: same data each time is fine -- more consistent on benchmarks

File: src/accountsdb/swiss_map.zig [hunk 28]
```diff
@@ -608,21 +610,21 @@ fn generateData(allocator: std.mem.Allocator, n_accounts: usize) !struct {
...
 
-    const accounts = try allocator.alloc(accounts_db.index.AccountRef, n_accounts);
+    const account_refs = try allocator.alloc(accounts_db.index.AccountRef, n_accounts);
     const pubkeys = try allocator.alloc(sig.core.Pubkey, n_accounts);
     for (0..n_accounts) |i| {
         random.bytes(&pubkeys[i].data);
-        accounts[i] = accounts_db.index.AccountRef.default();
-        accounts[i].pubkey = pubkeys[i];
+        account_refs[i] = accounts_db.index.AccountRef.default();
+        account_refs[i].pubkey = pubkeys[i];
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/swiss_map.zig]: [f79a4b2](https://github.com/Syndica/sig/pull/316/commits/f79a4b2c1a095f195343d5174620169d780d734f)

File: src/accountsdb/download.zig [hunk 29]
```diff
@@ -302,81 +304,97 @@ pub fn downloadSnapshotsFromGossip(
...
         min_mb_per_second: ?usize,
     ) !Self {
-        const file = try output_dir.createFile(filename, .{ .read = true });
-        defer file.close();
+        const file = try output_dir.createFile(filename, .{});
 
         // resize the file
         try file.seekTo(download_size - 1);
         _ = try file.write(&[_]u8{1});
         try file.seekTo(0);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/download.zig]: [f79a4b2](https://github.com/Syndica/sig/pull/316/commits/f79a4b2c1a095f195343d5174620169d780d734f)

File: src/benchmarks.zig (hunk 12 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 706be88d0c5d07fcc9804caa954076625dba552f

File: src/utils/fmt.zig (hunk 20 above)
Comment (0xNineteen) [reply src/utils/fmt.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8

File: src/benchmarks.zig (hunk 18 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8

File: src/benchmarks.zig (hunk 16 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8 - not sure, agreed should remove

File: src/accountsdb/db.zig (hunk 22 above)
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:512]: not sure what you mean - on `AccountFile.init` the `accounts_file` is never stored so a defer close is okay

File: src/benchmarks.zig (hunk 15 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8

File: src/benchmarks.zig (hunk 14 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8

File: src/benchmarks.zig (hunk 13 above)
Comment (0xNineteen) [reply src/benchmarks.zig]: 8948a0cb3b14df8d6d5d65356aefa394a1c27ee8

File: src/accountsdb/db.zig [hunk 30]
```diff
@@ -327,12 +326,10 @@ pub const AccountsDB = struct {
...
-        var loading_threads = try ArrayList(AccountsDB).initCapacity(
-            self.allocator,
-            n_parse_threads,
-        );
-        for (0..n_parse_threads) |_| {
-            var thread_db = loading_threads.addOneAssumeCapacity();
+
+        const loading_threads = try self.allocator.alloc(AccountsDB, n_parse_threads);
+        defer self.allocator.free(loading_threads);
+        for (loading_threads) |*thread_db| {
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:332]: think this way is better tbh -- see comment in your PR

[the remark this answers is not in the document]
Comment (InKryption) [reply src/accountsdb/db.zig:332]: The problem with this way is that it leaks resources in the error path.

File: src/accountsdb/db.zig (hunk 22 above)
[the remark this answers is not in the document]
Comment (InKryption) [reply src/accountsdb/db.zig:512]: Oh I'm way out of the loop then, I thought we still had the file handle, my bad.

File: src/accountsdb/db.zig (hunk 30 above)
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig:332]: makes sense - should be good to follow your approach - should be an easy merge conflict to fix / force-push over for whatever pr gets merged first

**Review by InKryption — APPROVED** [review]
LGTM

**Review by Sobeston — APPROVED** [review]
Gave it a try earlier, works great (& great work). No more nits, happy to merge.
